### PR TITLE
Rewrite in rust

### LINKED ISSUE #196: high cpu usage on server
The following command connects to catris running on localhost without `--lobbies`, enters name `asd`, selects ring mode from the menu, moves to the left enough to pass the middle square, and then makes the block go fast:

```
echo $'asd\r\nr\r\naaaaaaaas' | nc localhost 12345 >/dev/null
```

This causes about 25% cpu usage on my computer, or 30% with #195. Because of the `>/dev/null` at the end, the game isn't being displayed, so this is not just my terminal being slow. This would be pretty bad if my server ran many games simultaneously.

### LINKED ISSUE #212: seek in high scores code "produces undefined behaviour"
https://github.com/Akuli/catris/blob/64cf9b8f6bf093876d7e77fc35bebe17275ac9f1/catris/high_scores.py#L62

[The docs](https://docs.python.org/3/library/io.html?highlight=undefined%20behaviour#io.TextIOBase.seek) say that calling `.seek()` in text mode with anything other than zero or a return value of `.tell()` "produces undefined behaviour". It works in practice though.

### PR DESCRIPTION
**Pull request by Akuli**
Fixes #196 
Resolves #212

### CODE REVIEW

File: .github/workflows/github_actions.yml [hunk 1]
```diff
@@ -5,49 +5,28 @@ on:
...
-      - run: pip install -r requirements-dev.txt
-      - run: mypy --python-version 3.7 --strict catris/
-      - run: mypy --python-version 3.8 --strict catris/
-      - run: mypy --python-version 3.9 --strict catris/
-      - run: mypy --python-version 3.10 --strict catris/
-  pyflakes:
+    - uses: actions/checkout@v2
+    - uses: ATiltedTree/setup-rust@v1
+      with:
+        rust-version: stable
```
Comment (Akuli) [inline .github/workflows/github_actions.yml:14]: Should figure out if I can be more specific than this

File: .github/workflows/github_actions.yml [hunk 2]
```diff
@@ -5,49 +5,28 @@ on:
...
+    - run: cargo fmt --check
   eslint:
     runs-on: ubuntu-latest
     steps:
+    - uses: ATiltedTree/setup-rust@v1
+      with:
+        rust-version: stable
```
Comment (Akuli) [inline .github/workflows/github_actions.yml:29]: why is this in eslint section :DDDD

File: Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+debug = true
+
+[dependencies]
+tokio = { version = "1", features = ["full"] }
+weak-table = "*"
+rand = "*"
+lazy_static = "*"
+tokio-tungstenite = "*"
+# futures-util copied from tokio-tungstenite's Cargo.toml
+futures-util = { version = "0.3", default-features = false, features = ["sink", "std"] }
```
Comment (Akuli) [inline Cargo.toml:18]: Not sure if this is really needed?

File: README.md [hunk 4]
```diff
@@ -37,42 +37,60 @@ Please create an issue if this is a problem for you.
 
 ## Development
 
-If you're on Windows, use `py` instead of `python3` and `env\Scripts\activate` instead of `source env/bin/activate` below.
+You need to install rust. Just google for some instructions.
```
Comment (Akuli) [inline README.md:40]:
```suggestion
You need to install rust (the compiler, not the game). Just google for some instructions.
```

File: README.md [hunk 5]
```diff
@@ -37,42 +37,60 @@ Please create an issue if this is a problem for you.
...
 You can then open `http://localhost:8000/` in your web browser.
 
-Linters and formatters:
+I use `cargo fmt` to format my code. GitHub Actions ensures that it was used.
+
+
+## Deploying
+
+These instructions are mostly for me.
+If you want to run catris in a local network, see [local-playing.md](local-playing.md).
```
Comment (Akuli) [inline README.md:78]: Should check if link works

File: README.md [hunk 6]
```diff
@@ -37,42 +37,60 @@ Please create an issue if this is a problem for you.
...
+
+If you changed rust code, build the executable and copy it to the server:
 
 ```
-black catris        # Formats the code
-isort catris        # Formats and sorts imports
-mypy catris         # Type checker, detects common mistakes
-pyflakes catris     # Linter, detects some less common mistakes
+$ cargo build --release
+$ scp target/release/catris catris:/home/catris/catris
```
Comment (Akuli) [inline README.md:89]:
```suggestion
$ scp target/release/catris catris:/home/catris/catris
$ ssh catris
$ sudo systemctl restart catris
```

File: README.md [hunk 7]
```diff
@@ -37,42 +37,60 @@ Please create an issue if this is a problem for you.
...
+
+These instructions are mostly for me.
+If you want to run catris in a local network, see [local-playing.md](local-playing.md).
+If you run a bigger catris server, please let me know by creating an issue :)
+
+Look at `journalctl -fu catris` to make sure nobody is currently playing.
+If in the future there is always someone playing,
+use `/home/catris/catris_motd.txt` to clearly announce the downtime beforehand.
+
+If you changed rust code, build the executable and copy it to the server:
```
Comment (Akuli) [inline README.md:85]:
```suggestion
If you changed rust code, build the executable, copy it to the server, and restart the systemd service:
```

File: local-playing.md [hunk 8]
```diff
@@ -7,48 +7,11 @@ with several computers connected to the same router with Ethernet cables.
...
-```
-
-If you want to use the web UI, you also need to start a web server in a second terminal.
-The server must be started from the `web-ui` directory.
-
-```
-cd web-ui
-python3 -m http.server
-```
+Install rust and run the server as explained in
```
Comment (Akuli) [inline local-playing.md:12]:
```suggestion
Install rust and run the catris server with `cargo` as explained in
```

File: local-playing.md [hunk 9]
```diff
@@ -7,48 +7,11 @@ with several computers connected to the same router with Ethernet cables.
...
+[the development section of the README](README.md#development).
```
Comment (Akuli) [inline local-playing.md:13]: Should check if link works

File: Cargo.toml (hunk 3 above)
Comment (Akuli) [reply Cargo.toml:18]: it is

**Comment by Akuli** [comment]
Note for myself when I look at this later: The first rust commit was on `Sat May 21 16:11:44 2022 +0300` so it took me almost a month to get this rewritten in rust.

**Comment by Moosems** [comment]
This might be the biggest PR I've ever seen.

**Pull request merged**